cli/lib/kontena/cli/stacks/upgrade_command.rb

```diff
@@ -19,152 +22,197 @@ class UpgradeCommand < Kontena::Command
 
     option '--force', :flag, 'Force upgrade'
     option '--skip-dependencies', :flag, "Do not install any stack dependencies"
-    option '--dry-run', :flag, "Do not perform any uploading", hidden: true
+    option '--dry-run', :flag, "Simulate upgrade"
 
     requires_current_master
     requires_current_master_token
 
-    def normalize_local_data(stack_data, parent_name)
-      return nil if stack_data.nil? || stack_data.empty?
+    attr_reader :old_data, :new_data, :changes
 
-      depends = stack_data.delete('depends') || []
-      normalized_data = {
-        parent_name => stack_data.merge(
-          :loader => loader_class.for(stack_data['stack'])
-        )
-      }
+    def execute
+      set_env_variables(stack_name, current_grid)
 
-      depends.each do |stack|
-        key = "#{parent_name}-#{stack['name']}"
-        normalized_data.merge!(normalize_local_data(stack.merge('parent_name' => parent_name), key))
-      end
-      normalized_data
-    end
+      gather_data
```

Comment:
The use of `@old_data` and `@new_data` as pseudo-global variables bothers me. Why can't this method return something? It's so hard to follow the flow of data across these functions :confused:

Reply:
That's kind of what instance variables are for. Methods that do nothing to/with instance state are marked by [Reek](https://github.com/troessner/reek) as [Utility Functions](https://github.com/troessner/reek/blob/master/docs/Utility-Function.md) and are considered a code smell.